devito/types/sparse.py

```diff
@@ -128,21 +124,25 @@ def __subfunc_setup__(self, key, suffix, dtype=None):
             shape = (self.npoint, self.grid.dim)
 
         # Check if already a SubFunction
+        d = self.indices[self._sparse_position]
         if isinstance(key, SubFunction):
-            # Need to rebuild so the dimensions match the parent SparseFunction
-            indices = (self.indices[self._sparse_position], *key.indices[1:])
-            return key._rebuild(*indices, name=name, shape=shape,
-                                alias=self.alias, halo=None)
-        elif key is not None and not isinstance(key, Iterable):
-            raise ValueError("`%s` must be either SubFunction "
-                             "or iterable (e.g., list, np.ndarray)" % key)
+            if d in key.dimensions and not self.alias:
+                # From a reconstruction which leaves `dimensions` intact
+                return key
```

Comment:
For future reference, this is very problematic (both if and else case) because now `key`'s parent does not match the sparsefunction that uses it as a subfunction so can lead to very problematic runtime args/post_process so we may need to completely avoid this rebuild and always create a new subfunc using the key's data (need a '_local' initalizer